File: Source/OldCode/Archive/2005 TVA Code Library/Version/3.0/Local/Source/Tools/ConfigCrypter/ApplicationEvents.vb
Imports System.IO
Imports System.Reflection

Namespace My

    ' The following events are availble for MyApplication:
    ' 
    ' Startup: Raised when the application starts, before the startup form is created.
    ' Shutdown: Raised after all application forms are closed.  This event is not raised if the application terminates abnormally.
    ' UnhandledException: Raised if the application encounters an unhandled exception.
    ' StartupNextInstance: Raised when launching a single-instance application and the application is already active. 
    ' NetworkAvailabilityChanged: Raised when the network connection is connected or disconnected.
    Partial Friend Class MyApplication

        Private Sub MyApplication_Startup(ByVal sender As Object, ByVal e As Microsoft.VisualBasic.ApplicationServices.StartupEventArgs) Handles Me.Startup

            m_assemblyCache = New Dictionary(Of String, Assembly)
            LoadAssemblyFromResource("TVA.Core")

        End Sub

#Region " Load Embedded Assembly "

        Private m_addedResolver As Boolean
        Private m_assemblyCache As Dictionary(Of String, Assembly)

        Public Sub LoadAssemblyFromResource(ByVal assemblyName As String)

            ' Hook into assembly resolve event for current domain so we can load assembly from embedded resource
            If Not m_addedResolver Then
                AddHandler AppDomain.CurrentDomain.AssemblyResolve, AddressOf ResolveAssemblyFromResource
                m_addedResolver = True
            End If

            ' Load the assembly (this will invoke event that will resolve assembly from resource)
            AppDomain.CurrentDomain.Load(assemblyName)

        End Sub

        Private Function ResolveAssemblyFromResource(ByVal sender As Object, ByVal e As ResolveEventArgs) As System.Reflection.Assembly

            'LogMessage("Resolving assembly...")

            Dim resourceAssembly As System.Reflection.Assembly = Nothing
            Dim shortName As String = e.Name.Split(","c)(0)

            If Not m_assemblyCache.TryGetValue(shortName, resourceAssembly) Then
                ' Loop through all of the resources in the executing assembly
                For Each name As String In Assembly.GetEntryAssembly.GetManifestResourceNames()
                    ' See if the embedded resource name matches assembly we are trying to load
                    If String.Compare(Path.GetFileNameWithoutExtension(name), Me.GetType().Assembly.GetExportedTypes(0).Namespace & "." & shortName, True) = 0 Then
                        ' If so, load embedded resource assembly into a binary buffer
                        With Assembly.GetEntryAssembly.GetManifestResourceStream(name)
                            Dim length As Integer = CInt(.Length)
                            Dim buffer(length - 1) As Byte
                            .Read(buffer, 0, length)
                            .Close()

                            ' Load assembly from binary buffer
                            resourceAssembly = Assembly.Load(buffer)
                            m_assemblyCache.Add(shortName, resourceAssembly)
                            Exit For
                        End With
                    End If
                Next
            End If

            Return resourceAssembly

        End Function

#End Region

    End Class

End Namespace

File: Source/OldCode/Archive/2008 TVA Code Library/Source/Tools/ConfigCrypter/ApplicationEvents.cs
using System.Diagnostics;
using System;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Reflection;


namespace ConfigCrypter
{
	namespace My
	{
		
		// The following events are availble for MyApplication:
		//
		// Startup: Raised when the application starts, before the startup form is created.
		// Shutdown: Raised after all application forms are closed.  This event is not raised if the application terminates abnormally.
		// UnhandledException: Raised if the application encounters an unhandled exception.
		// StartupNextInstance: Raised when launching a single-instance application and the application is already active.
		// NetworkAvailabilityChanged: Raised when the network connection is connected or disconnected.
		partial class MyApplication
		{
			
			
			private void MyApplication_Startup(object sender, Microsoft.VisualBasic.ApplicationServices.StartupEventArgs e)
			{
				
				m_assemblyCache = new Dictionary<string, Assembly>();
				LoadAssemblyFromResource("PCS.Core");
				
			}
			
			#region " Load Embedded Assembly "
			
			private bool m_addedResolver;
			private Dictionary<string, Assembly> m_assemblyCache;
			
			public void LoadAssemblyFromResource(string assemblyName)
			{
				
				// Hook into assembly resolve event for current domain so we can load assembly from embedded resource
				if (! m_addedResolver)
				{
					AppDomain.CurrentDomain.AssemblyResolve += new System.ResolveEventHandler(ResolveAssemblyFromResource);
					m_addedResolver = true;
				}
				
				// Load the assembly (this will invoke event that will resolve assembly from resource)
				AppDomain.CurrentDomain.Load(assemblyName);
				
			}
			
			private System.Reflection.Assembly ResolveAssemblyFromResource(object sender, ResolveEventArgs e)
			{
				
				//LogMessage("Resolving assembly...")
				
				System.Reflection.Assembly resourceAssembly = null;
				string shortName = e.Name.Split(',')[0];
				
				object temp_object = resourceAssembly;
				System.Reflection.Assembly temp_SystemReflectionAssembly =  (System.Reflection.Assembly) (temp_object);
				if (! m_assemblyCache.TryGetValue(shortName, out temp_SystemReflectionAssembly))
				{
					// Loop through all of the resources in the executing assembly
					foreach (string name in System.Reflection.Assembly.GetEntryAssembly().GetManifestResourceNames())
					{
						// See if the embedded resource name matches assembly we are trying to load
						if (string.Compare(Path.GetFileNameWithoutExtension(name), this.GetType().Assembly.GetExportedTypes()[0].Namespace + "." + shortName, true) == 0)
						{
							// If so, load embedded resource assembly into a binary buffer
							System.IO.Stream with_1 = Assembly.GetEntryAssembly().GetManifestResourceStream(name);
							int length = (int) with_1.Length;
							byte[] buffer = new byte[length];
							with_1.Read(buffer, 0, length);
							with_1.Close();
							
							// Load assembly from binary buffer
							resourceAssembly = Assembly.Load(buffer);
							m_assemblyCache.Add(shortName, resourceAssembly);
							break;
						}
					}
				}
				
				return resourceAssembly;
				
			}
			
			#endregion
			
		}
		
	}
	
	
}
